File: backend/app/schemas/lab_membership_request.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.lab_membership_request import LabMembershipRequestStatus
from app.schemas.laboratory import LaboratoryCreate


class LabMembershipRequestCreateExisting(BaseModel):
    laboratory_id: UUID


class LabMembershipRequestCreateNewLab(LaboratoryCreate):
    pass


class LabMembershipRequestRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    user: "UserRead"
    laboratory_id: UUID
    laboratory: "LaboratoryRead"
    status: LabMembershipRequestStatus
    reviewed_by_id: Optional[UUID] = None
    reviewed_by: Optional["UserRead"] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime


from app.schemas.laboratory import LaboratoryRead
from app.schemas.user import UserRead

File: frontend/src/api/types/labMembershipRequest.ts
import type { User } from './auth'
import type { Laboratory, LaboratoryCreate } from './laboratory'

export type LabMembershipRequestStatus = 'pending' | 'approved' | 'denied'

export interface LabMembershipRequest {
  id: string
  user_id: string
  user: User
  laboratory_id: string
  laboratory: Laboratory
  status: LabMembershipRequestStatus
  reviewed_by_id: string | null
  reviewed_by: User | null
  reviewed_at: string | null
  created_at: string
}

export interface JoinExistingLabPayload {
  laboratory_id: string
}

export type NewLabRequestPayload = LaboratoryCreate
